app/channels/editor_channel.rb

```diff
@@ -0,0 +1,59 @@
+# The Editor Channel is opened when a user navigates to the `Edit` page of an
+# Evidence, Issue, Note, or Card. The channel is closed when the user navigates
+# away from the `edit` page. When open the channel receives updates from the
+# editor and saves them as new revisions of the resource.
+class EditorChannel < ApplicationCable::Channel
+  include ProjectScopedChannels
+  include PaperTrailActivity
+
+  attr_accessor :resource
+
+  AUTOSAVE_EVENT = 'auto-save'.freeze
+
+  def subscribed
+    reject and return unless find_resource(params)
+
+    stream_for [current_user, current_project, resource]
+  end
+
+  def save(params)
+    resource.paper_trail_event = AUTOSAVE_EVENT
+
+    if resource.update_attributes parsed_params(params['data'])
+      track_activity(resource, AUTOSAVE_EVENT)
+
+      self.class.broadcast_to([current_user, current_project, resource], resource.reload.updated_at.to_i)
+    end
+  end
+
+  private
+
+  def find_resource(params)
+    return unless %w[evidence issue note card].include? params['resource_type']
+
+    if params['resource_type'] == 'card'
+      authorized_list_ids = List.where(board_id: current_project.boards.select(:id)).select(:id)
+      @resource ||= Card.find_by(id: params['resource_id'], list_id: authorized_list_ids)
+    else
+      @resource ||= current_project.send(params['resource_type'].pluralize).find_by id: params['resource_id']
+    end
+  end
+
+  def parsed_params(data)
+    wrapped_params = Rack::Utils.parse_nested_query data
+    ActionController::Parameters.
+      new(wrapped_params).
+      require(@params['resource_type']).
+      permit(permissable_params(@params['resource_type']))
+  end
+
+  def permissable_params(type)
+    case type
+    when 'evidence' then %i[author content issue_id node_id]
+    when 'issue' then %i[tag_list text]
```

Comment:
We can add `state` to the list for the issue states.

Reply:
Like:  `when 'issue' then %i[tag_list text state]` ?

Reply:
Isn't this duplicating the standard app's Strong parameters definitions? I think that's a slippery slope. We're going to end up modifying in one place and forgetting to do so in the other.

Is there a way to reuse the strong parameter definition from each controller?

Reply:
That was the first adverse feeling I had to it as well. So here's my current thought:
It's not a slippery slope. We have duplicated text, but the domain is entirely different. Strong Params exists to control the flow of data to an entry point. Remember this functionality was removed from `attr_accessbile` and placed into controllers for more finite control at the point of entry. WebSocket channels are not controllers, it is a different entry point into our application, and may or may not share commonalities with the controller entry points. We _could_ find some dirty trick to re-use the code from the controller but I think it's not needed. We already have multiple entry points, and each controls it's own flow. The WebApp has it's own strong params, the API also declares it's own strong params, and now web sockets. It definitely has potential to be missed in future updates, but that wasn't untrue before.

That said. I know where you're coming from, but I believe it's the wrong angle.
So as it stands I have no plans to change this.

Reply:
What about re-using the StrongParameters syntax with `#require` and `#permit`? That way if Rails evolves something will break here and we'll remember to evolve this endpoint too.

I'm concerned about the proliferation of non-standard entry points. The API is bad enough, and the chances of anyone considering / remembering the web sockets endpoint are slimer.

Reply:
You mean the `#require` and `#permit` on line 49 and 50? 😄 

But the plays into the comment on those lines, about it being hard to read. So I think moving the permissable_params case statement directly into a strong params usage might make it simpler to grok.

Reply:
I've refactored the channel for readability. We are still defining the allowed params in the channel, and continuing to use StrongParams. Overall the file should be easier to understand.